File: 1644. Lowest Common Ancestor of a Binary Tree II.java
// Medium
// DFS
// Time:O(n),Space:O(h)
// https://leetcode.cn/problems/lowest-common-ancestor-of-a-binary-tree-ii/

class Solution {
    // Use DFS to find p and q
    private boolean foundP = false, foundQ = false;

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode lca = dfs(root, p, q);
        return (foundP & foundQ) ? lca : null;
    }
    private TreeNode dfs(TreeNode node, TreeNode p, TreeNode q) {
        if (node == null) return null;
        TreeNode left = dfs(node.left, p, q), right = dfs(node.right, p, q);
        if (node == p) {
            foundP = true;
            return node;
        }
        if (node == q) {
            foundQ = true;
            return node;
        }
        if (left != null && right != null) return node;
        return left != null ? left : right;
    }
}

File: 1001-2000/1644. Lowest Common Ancestor of a Binary Tree II.py
# Time:O(n),Space:O(h)

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(
        self, root: "TreeNode", p: "TreeNode", q: "TreeNode"
    ) -> "TreeNode":
        self.found_p = False
        self.found_q = False

        def dfs(node):
            if not node:
                return None

            if node == p:
                self.found_p = True
            if node == q:
                self.found_q = True

            left = dfs(node.left)
            right = dfs(node.right)

            if node == p or node == q:
                return node

            if left and right:
                return node

            return left if left else right

        lca = dfs(root)

        if self.found_p and self.found_q:
            return lca

        return None
